components/builder-api-client/src/lib.rs

```diff
@@ -139,10 +142,45 @@ impl Client {
         let mut encoded = String::new();
         res.read_to_string(&mut encoded).map_err(Error::IO)?;
         debug!("Body: {:?}", encoded);
-        let rd: ReverseDependencies = serde_json::from_str(&encoded).map_err(Error::JSON)?;
+        let rd: ReverseDependencies = serde_json::from_str(&encoded).map_err(Error::Json)?;
         Ok(rd.rdeps.to_vec())
     }
 
+    /// Promote a job group to a channel
+    ///
+    /// # Failures
+    ///
+    /// * Remote API Server is not available
+    pub fn job_group_promote(
+        &self,
+        group_id: u64,
+        channel: &str,
+        token: &str,
+    ) -> Result<Vec<String>> {
+        debug!("Promoting job group {} to channel {}", group_id, channel);
+        let url = format!("jobs/group/{}/promote/{}", group_id, channel);
+        let res = self.add_authz(self.0.post(&url), token).send().map_err(
+            Error::HyperError,
+        )?;
+
+        if res.status != StatusCode::Ok {
+            return Err(err_from_response(res));
+        }
+
+        match decoded_response::<UnsuccessfulJobGroupPromote>(res)
```

Comment:
Reviving [this thread](https://github.com/habitat-sh/habitat/pull/2866#discussion_r132986922)... I feel like this response could be structured differently so that this code is not as confusing. Maybe instead of calling it `UnsuccessfulJobGroupPromote`, it's just a `JobGroupPromoteResult` (or whatever). If no projects failed to promote, yay, that's a successful promotion. If some failed, then we create an appropriate error. That way, the flow is a bit more clear, and you don't have things like errors-that-aren't-really-errors.